ostd/src/cpu/local/cpu_local.rs

```diff
@@ -0,0 +1,219 @@
+        $(
+            #[link_section = ".cpu_local"]
+            $(#[$attr])* $vis static $name: $crate::cpu::local::CpuLocal<$t> = {
+                let val = $init;
+                // SAFETY: The per-CPU variable instantiated is statically
+                // stored in the special `.cpu_local` section.
+                unsafe {
+                    $crate::cpu::local::CpuLocal::__new(val)
+                }
+            };
+        )*
+    };
+}
+
+/// CPU-local objects.
+///
+/// CPU-local objects are instanciated once per CPU core. They can be shared to
+/// other cores. In the context of a preemptible kernel task, when holding the
+/// reference to the inner object, the object is always the one in the original
+/// core (when the reference is created), no matter which core the code is
+/// currently running on.
+///
+/// For the difference between [`CpuLocal`] and [`super::CpuLocalCell`], see
+/// [`super`].
+pub struct CpuLocal<T: 'static>(T);
+
+#[cfg(debug_assertions)]
+use core::sync::atomic::Ordering;
+
+#[cfg(debug_assertions)]
+use super::IS_INITIALIZED;
+
+impl<T: 'static> CpuLocal<T> {
+    /// Creates a new CPU-local object.
+    ///
+    /// Please do not call this function directly. Instead, use the
+    /// `cpu_local!` macro.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure that the object initialized by this
+    /// function resides in the `.cpu_local` section. Otherwise the
+    /// behavior is undefined.
+    #[doc(hidden)]
+    pub const unsafe fn __new(val: T) -> Self {
+        Self(val)
+    }
+
+    /// Get access to the underlying value with IRQs disabled.
+    ///
+    /// By this method, you can borrow a reference to the underlying value
+    /// even if `T` is not `Sync`. Because that it is per-CPU and IRQs are
+    /// disabled, no other running tasks can access it.
+    pub fn borrow_irq_disabled(&'static self) -> CpuLocalDerefGuard<'_, T> {
+        CpuLocalDerefGuard {
+            cpu_local: self,
+            _guard: InnerGuard::Created(trap::disable_local()),
+        }
+    }
+
+    /// Get access to the underlying value with a provided guard.
+    ///
+    /// Similar to [`CpuLocal::borrow_irq_disabled`], but you can provide
+    /// a guard to disable IRQs if you already have one.
+    pub fn borrow_with<'a>(
+        &'static self,
+        guard: &'a DisabledLocalIrqGuard,
+    ) -> CpuLocalDerefGuard<'a, T> {
+        CpuLocalDerefGuard {
+            cpu_local: self,
+            _guard: InnerGuard::Provided(guard),
+        }
+    }
+
+    /// Get access to the underlying value through a raw pointer.
+    ///
+    /// This function calculates the virtual address of the CPU-local object
+    /// based on the CPU-local base address and the offset in the BSP.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the reference to `self` is static.
+    unsafe fn get(&self) -> *const T {
```

Comment:
This is aligned with [`UnsafeCell::get`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.get). I prefer `get` because it really calculates the address. `as_ptr` seems to just do a cast.